### PR TITLE
[SPARK-40936][SQL][TESTS] Refactor `AnalysisTest#assertAnalysisErrorClass` by reusing the `SparkFunSuite#checkError`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims to refactor  `AnalysisTest#assertAnalysisErrorClass` method by reusing the `checkError` method in `SparkFunSuite`.

On the other hand, the signature of `AnalysisTest#assertAnalysisErrorClass` method is changed from 

```
protected def assertAnalysisErrorClass(
      inputPlan: LogicalPlan,
      expectedErrorClass: String,
      expectedMessageParameters: Map[String, String],
      caseSensitive: Boolean = true,
      line: Int = -1,
      pos: Int = -1): Unit
``` 
to 

```
protected def assertAnalysisErrorClass(
      inputPlan: LogicalPlan,
      expectedErrorClass: String,
      expectedMessageParameters: Map[String, String],
      queryContext: Array[QueryContext] = Array.empty,
      caseSensitive: Boolean = true): Unit
```

Then when we need to use `queryContext` instead of `line + pos` for assertion

### Why are the changes needed?
`assertAnalysisErrorClass` and `checkError` does the same work.

### Does this PR introduce _any_ user-facing change?
No,  just for test

### How was this patch tested?

- Pass GitHub Actions

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala [hunk 1]
```diff
@@ -183,30 +185,28 @@ trait AnalysisTest extends PlanTest {
         analyzer.checkAnalysis(analyzer.execute(inputPlan))
       }
 
-      if (e.getErrorClass != expectedErrorClass ||
-          e.messageParameters != expectedMessageParameters ||
-          (line >= 0 && e.line.getOrElse(-1) != line) ||
-          (pos >= 0) && e.startPosition.getOrElse(-1) != pos) {
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala:189]: It seems that the last condition was wrong:

```
(pos >= 0) && e.startPosition.getOrElse(-1) != pos
```

Should be

```
(pos >= 0 && e.startPosition.getOrElse(-1) != pos)
```

**Comment by LuciferYang** [comment]
cc @MaxGekk @HyukjinKwon @dongjoon-hyun Think again, does this refactor look more simple?

**Comment by LuciferYang** [comment]
> assertAnalysisErrorClass

Sounds good, let me try.  Set this to draft first and will ping you when it can be reviewed @MaxGekk

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 2]
```diff
@@ -716,9 +714,8 @@ class AnalysisSuite extends AnalysisTest with Matchers {
     assertAnalysisErrorClass(parsePlan("WITH t(x) AS (SELECT 1) SELECT * FROM t WHERE y = 1"),
       "UNRESOLVED_COLUMN.WITH_SUGGESTION",
       Map("objectName" -> "`y`", "proposal" -> "`t`.`x`"),
-      caseSensitive = true,
-      line = -1,
-      pos = -1)
+      Array(ExpectedContext("y", 46, 46))
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:717]: https://github.com/apache/spark/blob/cf086b10de784fc92ae8b4d16065823ace520a7a/core/src/test/scala/org/apache/spark/SparkFunSuite.scala#L319-L332

This change is due to the `checkError` method will perform a forced check when `actualQueryContext` is not empty. If we can to relax some check conditions,  can add a precondition `queryContext.nonEmpty` for the `queryContext` check.

**Comment by LuciferYang** [comment]
@MaxGekk 

Refactor `assertAnalysisErrorClass` method:

- Reuse `checkError` in `assertAnalysisErrorClass` 
- Use `queryContext` instead of `line + pos` in `assertAnalysisErrorClass`  method  signature
- Fixed related tests

Please review this  if you have time, thanks ~

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
Waiting for Ci.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @LuciferYang.

**Pull request closed without merging**